Assets/Scripts/Foliage/SpawnFoliage.cs

```diff
@@ -1,161 +1,247 @@
-    //[SerializeField] private Object[] grass = new Object[2];
-    [SerializeField] private Object[] stones = new Object[2];
-
-    // Start is called before the first frame update
-    void Start()
-    {
-        Random.InitState(seed);
+{
+    const int prefabLimit = 10000;
 
-        plantStones();
-        plantBushes();
-        plantTrees();
-    }
+    [Header("Trees")]
+    [SerializeField] private int treeLimit = 1000;
+    [SerializeField] private float treeLine = 25;
+    [SerializeField] private float treeAngleLimit = 30;
+    [SerializeField] private GameObject[] treePrefabs = new GameObject[4];
 
-    /// <summary>
-    /// Plants trees
-    /// </summary>
-    private void plantTrees()
+    [Header("Bushes")]
+    [SerializeField] private int bushLimit = 5000;
+    [SerializeField] private float bushLine = 35;
+    [SerializeField] private float bushAngleLimit = 30;
+    [SerializeField] private GameObject[] bushPrefab = new GameObject[4];
+
+    [Header("Rocks")]
+    [SerializeField] private int stoneLimit = 400;
+    [SerializeField] private float stoneLine = 36;
+    [SerializeField] private float stoneAngleLimit = 30;
+    [SerializeField] private GameObject[] stonePrefab = new GameObject[4];
+
+    [Header("Misc")]
+    [SerializeField] private int seed = 0;
+    [SerializeField] private bool DEBUG = false;
+
+
+    private int treeIndex = 0;
+    private int treeSpawnIndex = 0;
+    private Vector3[] treePositions = new Vector3[prefabLimit];
+
+    private int bushIndex = 0;
+    private int bushSpawnIndex = 0;
+    private Vector3[] bushPositions = new Vector3[prefabLimit];
+
+    private int stoneIndex = 0;
+    private int stoneSpawnIndex = 0;
+    private Vector3[] stonePositions = new Vector3[prefabLimit];
+
+
+    // Private members
+    private GameObject foliageHandler;
+
+    private Vector3 planetCenter;
+    private Planet planet;
+    private float planetRadius;
+    private float waterLevel;
+
+    void Update()
     {
-        Vector3 position;
-        RaycastHit hit;
 
-        // Loops thought all tree locations
-        for (int i = 0; i < treeLimit; i++)
+        if(planet != null && foliageHandler != null)
         {
-            // Creates a ball around the planet which shoots rays to check if it's legal to spawn foliage
-            position = Random.onUnitSphere * rayBallRadius;
-
-            if (Physics.Raycast(position, -position, out hit, rayLenght))
+            GameObject[] cameras = GameObject.FindGameObjectsWithTag("MainCamera");
+            for(int i = 0; i < cameras.Length; i++)
             {
-                if (hit.transform.tag == "Foliage" ||                         // Checks if it doesnt hit ground
-                    Mathf.Abs(Vector3.Angle(position, hit.normal)) > 30 ||    // Checks the angle
-                    hit.distance < (rayBallRadius - planetRadius) - treeLine)        // Checks the height
+                // Tries to spawn 100 of each every frame we are near the planet
```

Comment:
From you Update loop, it seems like you are casting rays every frame the camera is close to the planet. I don't see why this is necessary and it is costing performance. Should it not be fine to just send them once to initialize what position and rotation the objects can be placed in and then you can use that every time the player comes back to the player?
![image](https://user-images.githubusercontent.com/47697397/220206470-d46992e0-4496-449b-bd91-9742ddf05ab0.png)